{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception in user code:\n",
      "------------------------------------------------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\guobaba\\Anaconda3\\lib\\site-packages\\urllib3\\connection.py\", line 141, in _new_conn\n",
      "    (self.host, self.port), self.timeout, **extra_kw)\n",
      "  File \"C:\\Users\\guobaba\\Anaconda3\\lib\\site-packages\\urllib3\\util\\connection.py\", line 83, in create_connection\n",
      "    raise err\n",
      "  File \"C:\\Users\\guobaba\\Anaconda3\\lib\\site-packages\\urllib3\\util\\connection.py\", line 73, in create_connection\n",
      "    sock.connect(sa)\n",
      "ConnectionRefusedError: [WinError 10061] 由于目标计算机积极拒绝，无法连接。\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\guobaba\\Anaconda3\\lib\\site-packages\\urllib3\\connectionpool.py\", line 601, in urlopen\n",
      "    chunked=chunked)\n",
      "  File \"C:\\Users\\guobaba\\Anaconda3\\lib\\site-packages\\urllib3\\connectionpool.py\", line 357, in _make_request\n",
      "    conn.request(method, url, **httplib_request_kw)\n",
      "  File \"C:\\Users\\guobaba\\Anaconda3\\lib\\http\\client.py\", line 1239, in request\n",
      "    self._send_request(method, url, body, headers, encode_chunked)\n",
      "  File \"C:\\Users\\guobaba\\Anaconda3\\lib\\http\\client.py\", line 1285, in _send_request\n",
      "    self.endheaders(body, encode_chunked=encode_chunked)\n",
      "  File \"C:\\Users\\guobaba\\Anaconda3\\lib\\http\\client.py\", line 1234, in endheaders\n",
      "    self._send_output(message_body, encode_chunked=encode_chunked)\n",
      "  File \"C:\\Users\\guobaba\\Anaconda3\\lib\\http\\client.py\", line 1026, in _send_output\n",
      "    self.send(msg)\n",
      "  File \"C:\\Users\\guobaba\\Anaconda3\\lib\\http\\client.py\", line 964, in send\n",
      "    self.connect()\n",
      "  File \"C:\\Users\\guobaba\\Anaconda3\\lib\\site-packages\\urllib3\\connection.py\", line 166, in connect\n",
      "    conn = self._new_conn()\n",
      "  File \"C:\\Users\\guobaba\\Anaconda3\\lib\\site-packages\\urllib3\\connection.py\", line 150, in _new_conn\n",
      "    self, \"Failed to establish a new connection: %s\" % e)\n",
      "urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x000001A75328B128>: Failed to establish a new connection: [WinError 10061] 由于目标计算机积极拒绝，无法连接。\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\guobaba\\Anaconda3\\lib\\site-packages\\requests\\adapters.py\", line 440, in send\n",
      "    timeout=timeout\n",
      "  File \"C:\\Users\\guobaba\\Anaconda3\\lib\\site-packages\\urllib3\\connectionpool.py\", line 639, in urlopen\n",
      "    _stacktrace=sys.exc_info()[2])\n",
      "  File \"C:\\Users\\guobaba\\Anaconda3\\lib\\site-packages\\urllib3\\util\\retry.py\", line 388, in increment\n",
      "    raise MaxRetryError(_pool, url, error or ResponseError(cause))\n",
      "urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='localhost', port=8097): Max retries exceeded with url: /env/main (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x000001A75328B128>: Failed to establish a new connection: [WinError 10061] 由于目标计算机积极拒绝，无法连接。',))\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\guobaba\\Anaconda3\\lib\\site-packages\\visdom\\__init__.py\", line 446, in _send\n",
      "    data=json.dumps(msg),\n",
      "  File \"C:\\Users\\guobaba\\Anaconda3\\lib\\site-packages\\requests\\api.py\", line 112, in post\n",
      "    return request('post', url, data=data, json=json, **kwargs)\n",
      "  File \"C:\\Users\\guobaba\\Anaconda3\\lib\\site-packages\\requests\\api.py\", line 58, in request\n",
      "    return session.request(method=method, url=url, **kwargs)\n",
      "  File \"C:\\Users\\guobaba\\Anaconda3\\lib\\site-packages\\requests\\sessions.py\", line 508, in request\n",
      "    resp = self.send(prep, **send_kwargs)\n",
      "  File \"C:\\Users\\guobaba\\Anaconda3\\lib\\site-packages\\requests\\sessions.py\", line 618, in send\n",
      "    r = adapter.send(request, **kwargs)\n",
      "  File \"C:\\Users\\guobaba\\Anaconda3\\lib\\site-packages\\requests\\adapters.py\", line 508, in send\n",
      "    raise ConnectionError(e, request=request)\n",
      "requests.exceptions.ConnectionError: HTTPConnectionPool(host='localhost', port=8097): Max retries exceeded with url: /env/main (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x000001A75328B128>: Failed to establish a new connection: [WinError 10061] 由于目标计算机积极拒绝，无法连接。',))\n",
      "2018-09-07 14:03:09,783 - visdom - INFO - Socket refused connection, running socketless\n",
      "2018-09-07 14:03:10,886 - __main__ - INFO - Start ......\n",
      "2018-09-07 14:03:10,890 - __main__ - INFO - Create model ......\n",
      "2018-09-07 14:03:10,922 - __main__ - INFO - Load train set ......\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "===================================================\n",
      "ClassificationModel(\n",
      "  (qry_encode): VariableLengthGRU(\n",
      "    (gru): GRU(3, 5, num_layers=2, batch_first=True, bidirectional=True)\n",
      "  )\n",
      "  (qry_self_attention): Attention(\n",
      "    (w_qs): Linear(in_features=10, out_features=10, bias=True)\n",
      "    (w_ks): Linear(in_features=10, out_features=10, bias=True)\n",
      "    (w_vs): Linear(in_features=10, out_features=10, bias=True)\n",
      "    (attn): DotProductAttention()\n",
      "  )\n",
      "  (doc_encode): VariableLengthGRU(\n",
      "    (gru): GRU(23, 5, num_layers=2, batch_first=True, bidirectional=True)\n",
      "  )\n",
      "  (doc_self_attention): Attention(\n",
      "    (w_qs): Linear(in_features=10, out_features=10, bias=True)\n",
      "    (w_ks): Linear(in_features=10, out_features=10, bias=True)\n",
      "    (w_vs): Linear(in_features=10, out_features=10, bias=True)\n",
      "    (attn): DotProductAttention()\n",
      "  )\n",
      "  (dense_layer1): Linear(in_features=40, out_features=5, bias=True)\n",
      "  (out_layer): Linear(in_features=5, out_features=3, bias=True)\n",
      ")\n",
      "===================================================\n"
     ]
    }
   ],
   "source": [
    "import torch\n",
    "import visdom\n",
    "import logging\n",
    "import numpy as np\n",
    "import torch.nn as nn\n",
    "from torch.nn import utils\n",
    "import torch.optim as optim\n",
    "from model.models import ClassificationModel\n",
    "from utils.dataset import *\n",
    "\n",
    "logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(name)s - %(levelname)s - %(message)s')\n",
    "logger = logging.getLogger(__name__)\n",
    "\n",
    "# embed_mat = np.load('./mrc_data/vectors.npy')\n",
    "# embed = nn.Embedding.from_pretrained(torch.Tensor(embed_mat))\n",
    "# embed = nn.Embedding.from_pretrained(torch.randn([3,3]))\n",
    "embed = nn.Embedding.from_pretrained(torch.randn([VOCAB.size(), 3]))\n",
    "embed_dim = 3\n",
    "hidden_dim = 5\n",
    "lr = 0.001\n",
    "batch_size = 32\n",
    "weight_decay = 0.0001\n",
    "\n",
    "vis = visdom.Visdom()\n",
    "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
    "\n",
    "\n",
    "def draw(ep, acc, title):\n",
    "    vis.line(X=torch.Tensor(1).fill_(ep).long(),\n",
    "             Y=torch.Tensor(1).fill_(acc).float(),\n",
    "             win=title,\n",
    "             update='append' if ep != 0 else None,\n",
    "             opts={'title': title, 'xlabel': 'EPOCH', 'ylabel': 'ACCURACY'})\n",
    "\n",
    "\n",
    "def evaluate(model, validset):\n",
    "    total = 0\n",
    "    correct_total = 0\n",
    "    model.eval()\n",
    "    for ii in range(validset.__len__()):\n",
    "        doc_pad, doc_lens, doc_mask, qry_pad, qry_lens, qry_mask, aws = validset[ii]\n",
    "        doc_pad = embed(doc_pad).to(device)\n",
    "        qry_pad = embed(qry_pad).to(device)\n",
    "        pred = model(doc_pad, doc_lens, doc_mask, qry_pad, qry_lens, qry_mask)\n",
    "        pred = torch.argmax(pred, 1)\n",
    "        correct_total += (pred == aws).sum().cpu().item()\n",
    "        total += doc_pad.size(0)\n",
    "        del doc_pad, doc_lens, doc_mask, qry_pad, qry_lens, qry_mask, aws\n",
    "\n",
    "    accuracy = correct_total / total\n",
    "    return accuracy\n",
    "\n",
    "\n",
    "def train(model: nn.Module, trainset: MRC_Dataset, epoch, validset: MRC_Dataset):\n",
    "\n",
    "    loss_func = nn.CrossEntropyLoss()\n",
    "\n",
    "    parameters = filter(lambda p: p.requires_grad, model.parameters())\n",
    "\n",
    "    optimizer = optim.Adam(parameters, lr=lr, weight_decay=weight_decay)\n",
    "\n",
    "    base_acc = 0.74\n",
    "    for ep in range(epoch):\n",
    "\n",
    "        model.train()\n",
    "\n",
    "        trainset.shuffle()\n",
    "\n",
    "        total = 0\n",
    "        correct_total = 0\n",
    "\n",
    "        for ii in range(trainset.__len__()):\n",
    "\n",
    "            optimizer.zero_grad()\n",
    "\n",
    "            doc_pad, doc_lens, doc_mask, qry_pad, qry_lens, qry_mask, aws = trainset[ii]\n",
    "\n",
    "            doc_pad = embed(doc_pad).to(device)\n",
    "            qry_pad = embed(qry_pad).to(device)\n",
    "\n",
    "            pred = model(doc_pad, doc_lens, doc_mask, qry_pad, qry_lens, qry_mask)\n",
    "\n",
    "            loss = loss_func(pred, aws)\n",
    "\n",
    "            total += doc_pad.size(0)\n",
    "            pred = torch.argmax(pred, 1)\n",
    "            correct= (pred == aws).sum().cpu().item()\n",
    "            correct_total += correct\n",
    "\n",
    "            logger.info(\"ep: %d, loss: %f, correct: %d\" % (ep, loss.item(), correct))\n",
    "\n",
    "            if (ep * trainset.__len__() + ii + 1) % 10 == 0:\n",
    "                vis.line(X=torch.Tensor(1).fill_(ep * trainset.__len__() + ii + 1).long(),\n",
    "                         Y=torch.Tensor(1).fill_(loss.item()).float(),\n",
    "                         win=\"Loss\",\n",
    "                         update='append' if ep != 0 or ii != 0 else None,\n",
    "                         opts={'title': 'Loss', 'xlabel': 'Times', 'ylabel': 'Loss'})\n",
    "\n",
    "            loss.backward()\n",
    "\n",
    "            utils.clip_grad_value_(parameters, 5.0)\n",
    "\n",
    "            optimizer.step()\n",
    "\n",
    "            del doc_pad, doc_lens, doc_mask, qry_pad, qry_lens, qry_mask, loss\n",
    "\n",
    "        train_accuracy = correct_total / total\n",
    "        draw(ep, train_accuracy, \"Train ACC\")\n",
    "\n",
    "        valid_accuracy = evaluate(model, validset)\n",
    "        draw(ep, valid_accuracy, \"Valid ACC\")\n",
    "\n",
    "        logger.info(\"ep: %d, train accuracy: %f, valid accuracy: %f\" % (ep, train_accuracy, valid_accuracy))\n",
    "\n",
    "        if valid_accuracy > base_acc:\n",
    "            base_acc = valid_accuracy\n",
    "            torch.save(model.state_dict(),\n",
    "                       '%s_acc-%f_ep-%d.pth' % ('checkpoints/c_m', valid_accuracy, ep))\n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "\n",
    "    logger.info(\"Start ......\")\n",
    "\n",
    "    logger.info(\"Create model ......\")\n",
    "    c_model = ClassificationModel(embed_dim, hidden_dim).to(device)\n",
    "    print(\"===================================================\")\n",
    "    print(c_model)\n",
    "    print(\"===================================================\")\n",
    "    \n",
    "\n",
    "    logger.info(\"Load train set ......\")\n",
    "    train_set = MRC_Dataset(batch_size=batch_size, is_trainset=False)\n",
    "\n",
    "    logger.info(\"Load valid set ......\")\n",
    "    #valid_set = MRC_Dataset(batch_size=batch_size, is_trainset=False)\n",
    "\n",
    "    train(c_model, train_set, 50, train_set)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'Visdom' object has no attribute 'updateTrace'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-9-2f2292483899>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m     12\u001b[0m \u001b[0mx\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mt\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0marange\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;36m9\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;36m0.1\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     13\u001b[0m \u001b[0my\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mx\u001b[0m\u001b[1;33m**\u001b[0m\u001b[1;36m2\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m/\u001b[0m\u001b[1;36m9\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 14\u001b[1;33m \u001b[0mvis\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mupdateTrace\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mX\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mx\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mY\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0my\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mwin\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'www'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mname\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'zhe ci de '\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mAttributeError\u001b[0m: 'Visdom' object has no attribute 'updateTrace'"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[WinError 10054] 远程主机强迫关闭了一个现有的连接。\n",
      "[WinError 10054] 远程主机强迫关闭了一个现有的连接。\n",
      "[WinError 10054] 远程主机强迫关闭了一个现有的连接。\n",
      "[WinError 10054] 远程主机强迫关闭了一个现有的连接。\n",
      "[WinError 10054] 远程主机强迫关闭了一个现有的连接。\n",
      "[WinError 10054] 远程主机强迫关闭了一个现有的连接。\n"
     ]
    }
   ],
   "source": [
    "import visdom\n",
    "import torch as t\n",
    "vis = visdom.Visdom(env=u'test1')\n",
    "x= t.arange(1,30,0.01)\n",
    "y=t.sin(x)\n",
    "vis.line(X=x,Y=y,win='sinx',opts={'title':'y=sin(x)'})\n",
    "for ii in range(0,1000):\n",
    "    x= t.Tensor([ii])\n",
    "    y=x\n",
    "    vis.line(X=x,Y=y,win='ppp',update='append' if ii >0 else None)\n",
    "    \n",
    "x=t.arange(0,9,0.1)\n",
    "y=(x**2)/9\n",
    "vis.updateTrace(X=x,Y=y,win='www',name='zhe ci de ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
